### ISSUE TITLE
new dataset and some questions :)

### ISSUE THREAD

**Issue by Aml-Hassan-Abd-El-hamid** [ml-unito/federation_boosting#1]
Hi there,

Thank you very much for sharing the code for your paper, your work is extremely interesting.

I want to try this code on another dataset though so how can I do that? I'm kinda lost in the code files and want to know how could I adjust the code to use it with my dataset.

**Comment by Aml-Hassan-Abd-El-hamid** [comment]
Also, I want to make sure of some things:

When you set the flag --non-iidness=uniform, do you mean that the distribution is iid?
In the image below, those results belong to different clients, right? if I'm right, do the results belong to the locally trained model or the federated global model? and why does the number of estimators differ between them?
![Screenshot from 2023-06-13 14-54-29](https://github.com/ml-unito/federation_boosting/assets/66205928/6e1f2d4c-1e44-47be-8307-6c78caf0311f)

**Comment by Aml-Hassan-Abd-El-hamid** [comment]
Also, the dataset that I'm dealing with is highly imbalanced is there is way to add a balanced Adaboost instead of a regular Adaboost?

**Comment by makgyver** [comment]
Hi,
I think the fastest way is to add the dataset in the `ijcnn_exps.py` file. 
Specifically, you need to modify the following parts:
- add to the `Datasets`enum your dataset;
- add a new if branch in `load_classification_dataset`, in which you handle the new dataset (loading and splitting).

Then, you can run the command
````
python ijcnn_exps.py run YOUR_DATASET [OPTIONS]
````

**Comment by makgyver** [comment]
> Also, I want to make sure of some things:
> 
> When you set the flag --non-iidness=uniform, do you mean that the distribution is iid? In the image below, those results belong to different clients, right? if I'm right, do the results belong to the locally trained model or the federated global model? and why does the number of estimators differ between them? ![Screenshot from 2023-06-13 14-54-29](https://user-images.githubusercontent.com/66205928/245466549-6e1f2d4c-1e44-47be-8307-6c78caf0311f.png)

No, the logger outputs the evaluation every 10 rounds (that is why you see the number of estimators increasing by 10), and, by default, it is the evaluation of the current global model on a separate test set (while the training error is there only for debugging purposes, it is the error overall the training set which is not possible to compute in a real federated setting)

**Comment by makgyver** [comment]
> Also, the dataset that I'm dealing with is highly imbalanced is there is way to add a balanced Adaboost instead of a regular Adaboost?

If you want to apply some kind of strategy to deal with imbalanced datasets, you should design something that can be computed client-side. In that case, you need to change how the client works.

Currently, the code (`fed_algorithms.py`) is not that friendly because there is no clear separation between client and server (although everything works as in an actual federated setting). However, we have already developed a new version of the code that follows the client/server implementation described in the paper, but we haven't released that yet. 

Given your interest, I will try to do it asap so that you can play with it.

**Comment by Aml-Hassan-Abd-El-hamid** [comment]
Hi @makgyver 

first of all, thank you very very much for the fast and helpful response.

1- I did as you said, added my dataset with its split code into the if conditions, and added its name to the class: Datasets, and I ran it using the following command 
``` python3 ijcnn_exps.py run ibm```
that's how I added my dataset to the if:
```
if name=="ibm":
        df=pd.read_csv("ibm_ada_ready.csv")
        y=df['Is Fraud?']
        X=df.drop(["Is Fraud?","Errors?"],axis=1)
        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 0, stratify=y)
        return X_train.to_numpy(),X_test.to_numpy(), y_train.to_numpy(), y_test.to_numpy()

```
and the logger output some crazy results, which you can see below:
![Screenshot from 2023-06-13 16-13-31](https://github.com/ml-unito/federation_boosting/assets/66205928/9e9630ad-b84f-4dab-ba1a-710540127896)

I tried again with different command : 
```
python3 ijcnn_exps.py run --seed=0 --n-clients=10  --non-iidness=uniform --tags=IJCNN --no-test-run  ibm
```
And it just kept going like this:

![Screenshot from 2023-06-13 16-14-43](https://github.com/ml-unito/federation_boosting/assets/66205928/4a691864-7904-4ac4-8302-b4a518371b90)

I'm calling those results crazy because, before trying your code I trained a regular sklearn Adaboost with 100 n_estimators on the same dataset and got the results below , Am I doing something wrong? Do I need to change any other parts of your code before testing it again?
![Screenshot from 2023-06-13 16-16-04](https://github.com/ml-unito/federation_boosting/assets/66205928/a2e7d36d-a01e-4f4a-ba2a-36c8c6aee54c)

**Comment by makgyver** [comment]
I think that the problem is the use of the "micro" instead of the "macro". Try the latter.

**Comment by Aml-Hassan-Abd-El-hamid** [comment]
yes, you were absolutely true, I changed them and now, I'm back to reality :) Thank you again for your quick help, I really appreciate it.
![Screenshot from 2023-06-13 16-25-02](https://github.com/ml-unito/federation_boosting/assets/66205928/b81748a3-ac76-4fdd-8a5b-30abc371ba2b)

**Comment by makgyver** [comment]
You are welcome!

**Comment by Aml-Hassan-Abd-El-hamid** [comment]
Hey @makgyver 

I just want to give you an update on the federated training results that I find really interesting, look at the results below:

![Screenshot from 2023-06-13 17-15-48](https://github.com/ml-unito/federation_boosting/assets/66205928/c9042acb-b4af-4f19-bc84-14e7f87ddcb5)


Although I was using AdaBoost with the whole dataset before, I wasn't able to get any better than the results shown below,  I have no clue why your federated version of  AdaBoost managed to get those results. I haven't modified it in any way, yet. 

![Screenshot from 2023-06-13 16-16-04](https://github.com/ml-unito/federation_boosting/assets/66205928/795774e4-102c-4a1e-9746-ee14a9a726d8)

**Comment by makgyver** [comment]
Hi, are you using the same weak classifier? Is the train/test split the same? Anyway, there are several "moving parts" that could cause those results.

**Comment by Aml-Hassan-Abd-El-hamid** [comment]
I'm using a decision tree as a weak classifier and the same train/test split, I'll try and dig more for the differences, but right now I can say that raising the number of n_estimators is the only thing that makes the original adadboost come closer to the performance of the federated adaboost. - raised from 100 to 1000-

**Comment by makgyver** [comment]
Yes, but it also depends on the hyper-parameters of the decision tree. I think that in our implementation, we have a maximum number of leaves set to 10. If you are using another setting, the results might differ. Otherwise, you should investigate what makes the federation better than the centralized version. Maybe it could result from a kind of "regularization effect" due to the data distribution across the clients. This last one is only a bold guess :)

**Comment by Aml-Hassan-Abd-El-hamid** [comment]
hey @makgyver 
You're right about the effect of the hyper-parameter of the tree, the way that you defined your decision tree is different than it's defined in the [sklearn implementation for adaboost](https://github.com/scikit-learn/scikit-learn/blob/364c77e04/sklearn/ensemble/_weight_boosting.py#L330) 
They define their tree with a max_depth=1 to make it a stump tree.
 Your tree definition doesn't specify the maximum depth which gives its nodes the ability to expand until all leaves are pure, also they don't specify the max_leaf_nodes.
I tried the code from [that article ](https://towardsdatascience.com/adaboost-from-scratch-37a936da3d50) after altering the definition of the decision tree to look like the one you used and managed to get those results:
![Screenshot from 2023-06-15 20-23-58](https://github.com/ml-unito/federation_boosting/assets/66205928/089bcd1b-3640-4a39-84fd-e9c87f376906)

**Comment by Aml-Hassan-Abd-El-hamid** [comment]
Hey @makgyver 
Can you please tell me what's the difference between the models: "same", "distsamme", "preweaksamme" and "adaboost.f1"?

**Comment by makgyver** [comment]
> Hey @makgyver Can you please tell me what's the difference between the models: "same", "distsamme", "preweaksamme" and "adaboost.f1"?

- Samme is a multi-class implementation of (centralized) adaboost (with 2 classes is exactly adaboost) 
- Distsamme, PreweakSamme and Adaboost.F1 are three federated versions of Samme. Generally, PreweakSamme and Adaboost.F1 perform better than Distsamme.

**Comment by Aml-Hassan-Abd-El-hamid** [comment]
Hey @makgyver 

I wanted to give you a quick update on the imbalance situation and ask you a couple of questions:
So I managed to make an imbalanced dataset by applying a simple trick, I redefined your weak learner to be a weighted decision tree:
``` 
weights = {0:1.0, 1:100.0}
WEAK_LEARNER = DecisionTreeClassifier(random_state=seed,max_leaf_nodes=10,class_weight=weights)
```
and I managed to get slightly better results than the one using my model which is trained on the whole dataset using the following command:
``` 
 python3 ijcnn_exps.py run --seed=0 --model="adaboost.f1" --n-clients=10  --non-iidness=uniform --tags=IJCNN --no-test-run  ibm
```
I got:
![Screenshot from 2023-06-17 02-54-23](https://github.com/ml-unito/federation_boosting/assets/66205928/8d8ea7b1-8ebe-4b6f-a1d4-21da83d439c0)

And just for reference that's the code that I'm using for adaboost:
```
#%% CREATE ADABOOST FROM SCRATCH
### Alvaro Corrales
### April 2021 (work in progress)

# Imports
import numpy as np
import pandas as pd
from sklearn.tree import DecisionTreeClassifier

# Helper functions
def compute_error(y, y_pred, w_i):
    '''
    Calculate the error rate of a weak classifier m. Arguments:
    y: actual target value
    y_pred: predicted value by weak classifier
    w_i: individual weights for each observation

    
    Note that all arrays should be the same length
    '''

    return (sum(w_i * (np.not_equal(y, y_pred)).astype(int)))/sum(w_i)

def compute_alpha(error):
    '''
    Calculate the weight of a weak classifier m in the majority vote of the final classifier. This is called
    alpha in chapter 10.1 of The Elements of Statistical Learning. Arguments:
    error: error rate from weak classifier m
    '''
    return np.log((1 - error) / error)

def update_weights(w_i, alpha, y, y_pred):
    ''' 
    Update individual weights w_i after a boosting iteration. Arguments:
    w_i: individual weights for each observation
    y: actual target value
    y_pred: predicted value by weak classifier  
    alpha: weight of weak classifier used to estimate y_pred
    '''  
    return w_i * np.exp(alpha * (np.not_equal(y, y_pred)).astype(int))

# Define AdaBoost class
class AdaBoost:
    
    def __init__(self):
        # self.w_i = None
        self.alphas = []
        self.G_M = []
        self.M = None
        self.training_errors = []
        self.prediction_errors = []

    def fit(self, X, y, M = 100):
        '''
        Fit model. Arguments:
        X: independent variables
        y: target variable
        M: number of boosting rounds. Default is 100
        '''
        
        # Clear before calling
        self.alphas = [] 
        self.training_errors = []
        self.M = M

        # Iterate over M weak classifiers
        for m in range(0, M):
            
            # Set weights for current boosting iteration
            if m == 0:
                w_i = np.ones(len(y)) * 1 / len(y)  # At m = 0, weights are all the same and equal to 1 / N
            else:
                w_i = update_weights(w_i, alpha_m, y, y_pred)
            # print(w_i)
            weights = {-1:1.0, 1:100.0}
            # (a) Fit weak classifier and predict labels
            #G_m = DecisionTreeClassifier(max_depth = 1)     # Stump: Two terminal-node classification tree
            G_m = DecisionTreeClassifier( max_leaf_nodes=10,class_weight=weights)
            G_m.fit(X, y, sample_weight = w_i)
            y_pred = G_m.predict(X)
            
            self.G_M.append(G_m) # Save to list of weak classifiers

            # (b) Compute error
            error_m = compute_error(y, y_pred, w_i)
            self.training_errors.append(error_m)
            # print(error_m)

            # (c) Compute alpha
            alpha_m = compute_alpha(error_m)
            self.alphas.append(alpha_m)
            # print(alpha_m)

        assert len(self.G_M) == len(self.alphas)


    def predict(self, X):
        '''
        Predict using fitted model. Arguments:
        X: independent variables
        '''

        # Initialise dataframe with weak predictions for each observation
        weak_preds = pd.DataFrame(index = range(len(X)), columns = range(self.M)) 

        # Predict class label for each weak classifier, weighted by alpha_m
        for m in range(self.M):
            y_pred_m = self.G_M[m].predict(X) * self.alphas[m]
            weak_preds.iloc[:,m] = y_pred_m

        # Estimate final predictions
        y_pred = (1 * np.sign(weak_preds.T.sum())).astype(int)

        return y_pred
      
    def error_rates(self, X, y):
        '''
        Get the error rates of each weak classifier. Arguments:
        X: independent variables
        y: target variables associated to X
        '''
        
        self.prediction_errors = [] # Clear before calling
        
        # Predict class label for each weak classifier
        for m in range(self.M):
            y_pred_m = self.G_M[m].predict(X)          
            error_m = compute_error(y = y, y_pred = y_pred_m, w_i = np.ones(len(y)))
            self.prediction_errors.append(error_m)
```

and that's the results that I got using it and the whole dataset:

![Screenshot from 2023-06-17 02-55-50](https://github.com/ml-unito/federation_boosting/assets/66205928/f5e61863-2b66-4579-9f1d-d6c5364b591a)


Now my question is:

How can I get some statistics on the distributed datasets that each client has?
I tried to inject some code on the function named: "distribute_dataset" thinking that it's the one responsible for dividing data,  but it doesn't seem like the code is using that function at all!

**Comment by Aml-Hassan-Abd-El-hamid** [comment]
Also, I conducted another experiment using another imbalanced dataset and weighted decision tree as a weak learner, and this time, the federated model did worse than the model trained with the whole dataset:
![Screenshot from 2023-06-17 03-59-25](https://github.com/ml-unito/federation_boosting/assets/66205928/0e9c2950-6654-4816-b3d5-6699aad883a6)
![Screenshot from 2023-06-17 04-00-55](https://github.com/ml-unito/federation_boosting/assets/66205928/58cc0268-b635-49d7-93e1-b9849bd424e6)

I made sure to test on the same train/test split for the federated model and the normal model. and I chose the no.of clients to be 4 as I thought that this dataset is too small to be divided into 10 clients.
